File: app/src/main/java/com/example/inventoryobat/MainViewModel.java
package com.example.inventoryobat;

import androidx.lifecycle.LiveData;
import androidx.lifecycle.MutableLiveData;
import androidx.lifecycle.ViewModel;

import java.util.ArrayList;
import java.util.List;

import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;

public class MainViewModel extends ViewModel {
    private final ApiService apiService;
    private final MutableLiveData<List<Obat>> obatListLiveData = new MutableLiveData<>();
    private final MutableLiveData<List<Supplier>> supplierListLiveData = new MutableLiveData<>();
    private final MutableLiveData<Boolean> loginStatus = new MutableLiveData<>();
    private final MutableLiveData<Boolean> isLoading = new MutableLiveData<>();
    private final MutableLiveData<String> errorMessage = new MutableLiveData<>();

    public MainViewModel() {
        apiService = ApiConfig.getApiService();
    }

    // ========== LOGIN ==========
    public LiveData<Boolean> getLoginStatus() {
        return loginStatus;
    }

    public void login(String username, String password) {
        // Untuk sekarang, validasi lokal (bisa dikembangkan ke server)
        if ("admin".equals(username) && "admin".equals(password)) {
            loginStatus.setValue(true);
        } else {
            loginStatus.setValue(false);
        }
    }

    // ========== OBAT ==========
    public LiveData<List<Obat>> getObatList() {
        return obatListLiveData;
    }

    public void loadAllObat() {
        isLoading.setValue(true);
        apiService.getAllObats().enqueue(new Callback<ApiResponse<List<Obat>>>() {
            @Override
            public void onResponse(Call<ApiResponse<List<Obat>>> call, Response<ApiResponse<List<Obat>>> response) {
                isLoading.setValue(false);
                if (response.isSuccessful() && response.body() != null) {
                    obatListLiveData.setValue(response.body().getData());
                } else {
                    errorMessage.setValue("Gagal memuat data obat");
                }
            }

            @Override
            public void onFailure(Call<ApiResponse<List<Obat>>> call, Throwable t) {
                isLoading.setValue(false);
                errorMessage.setValue("Error: " + t.getMessage());
            }
        });
    }

    public void loadObatByJenis(String jenis) {
        isLoading.setValue(true);
        apiService.getObatByJenis(jenis).enqueue(new Callback<ApiResponse<List<Obat>>>() {
            @Override
            public void onResponse(Call<ApiResponse<List<Obat>>> call, Response<ApiResponse<List<Obat>>> response) {
                isLoading.setValue(false);
                if (response.isSuccessful() && response.body() != null) {
                    obatListLiveData.setValue(response.body().getData());
                } else {
                    errorMessage.setValue("Gagal memuat data obat");
                }
            }

            @Override
            public void onFailure(Call<ApiResponse<List<Obat>>> call, Throwable t) {
                isLoading.setValue(false);
                errorMessage.setValue("Error: " + t.getMessage());
            }
        });
    }

    public LiveData<Obat> getObatById(int id) {
        MutableLiveData<Obat> result = new MutableLiveData<>();
        apiService.getObatById(id).enqueue(new Callback<ApiResponse<Obat>>() {
            @Override
            public void onResponse(Call<ApiResponse<Obat>> call, Response<ApiResponse<Obat>> response) {
                if (response.isSuccessful() && response.body() != null) {
                    result.setValue(response.body().getData());
                }
            }

            @Override
            public void onFailure(Call<ApiResponse<Obat>> call, Throwable t) {
                errorMessage.setValue("Error: " + t.getMessage());
            }
        });
        return result;
    }

    public void deleteObat(int id) {
        apiService.deleteObat(id).enqueue(new Callback<ApiResponse<String>>() {
            @Override
            public void onResponse(Call<ApiResponse<String>> call, Response<ApiResponse<String>> response) {
                if (response.isSuccessful()) {
                    loadAllObat();
                }
            }

            @Override
            public void onFailure(Call<ApiResponse<String>> call, Throwable t) {
                errorMessage.setValue("Error: " + t.getMessage());
            }
        });
    }

    public void updateStock(int id, int newStock) {
        apiService.updateStock(id, new UpdateStockRequest(newStock))
                .enqueue(new Callback<ApiResponse<Obat>>() {
                    @Override
                    public void onResponse(Call<ApiResponse<Obat>> call, Response<ApiResponse<Obat>> response) {
                        if (response.isSuccessful()) {
                            loadAllObat();
                        }
                    }

                    @Override
                    public void onFailure(Call<ApiResponse<Obat>> call, Throwable t) {
                        errorMessage.setValue("Error: " + t.getMessage());
                    }
                });
    }

    // ========== SUPPLIER ==========
    public LiveData<List<Supplier>> getSupplierList() {
        return supplierListLiveData;
    }

    public void loadAllSuppliers() {
        apiService.getAllSuppliers().enqueue(new Callback<ApiResponse<List<Supplier>>>() {
            @Override
            public void onResponse(Call<ApiResponse<List<Supplier>>> call, Response<ApiResponse<List<Supplier>>> response) {
                if (response.isSuccessful() && response.body() != null) {
                    supplierListLiveData.setValue(response.body().getData());
                }
            }

            @Override
            public void onFailure(Call<ApiResponse<List<Supplier>>> call, Throwable t) {
                errorMessage.setValue("Error: " + t.getMessage());
            }
        });
    }

    // ========== LOADING & ERROR ==========
    public LiveData<Boolean> getIsLoading() {
        return isLoading;
    }

    public LiveData<String> getErrorMessage() {
        return errorMessage;
    }
}

File: app/src/main/java/com/example/inventoryobat/MainViewModel.kt
package com.example.inventoryobat

import androidx.lifecycle.LiveData
import androidx.lifecycle.MutableLiveData
import androidx.lifecycle.ViewModel
import com.google.firebase.auth.FirebaseAuth
import retrofit2.Call
import retrofit2.Callback
import retrofit2.Response

class MainViewModel : ViewModel() {

    private val apiService: ApiService = ApiConfig.getApiService()
    private val mAuth: FirebaseAuth = FirebaseAuth.getInstance()

    private val _obatList = MutableLiveData<List<Obat>>()
    val obatList: LiveData<List<Obat>> get() = _obatList

    private val _supplierList = MutableLiveData<List<Supplier>>()
    val supplierList: LiveData<List<Supplier>> get() = _supplierList

    private val _loginStatus = MutableLiveData<Boolean>()
    val loginStatus: LiveData<Boolean> get() = _loginStatus

    private val _loginError = MutableLiveData<String>()
    val loginError: LiveData<String> get() = _loginError

    private val _isLoading = MutableLiveData<Boolean>()
    val isLoading: LiveData<Boolean> get() = _isLoading

    private val _errorMessage = MutableLiveData<String>()
    val errorMessage: LiveData<String> get() = _errorMessage

    fun login(email: String, password: String) {
        _isLoading.value = true
        mAuth.signInWithEmailAndPassword(email, password)
            .addOnCompleteListener { task ->
                _isLoading.value = false
                if (task.isSuccessful) {
                    _loginStatus.value = true
                } else {
                    _loginStatus.value = false
                    _loginError.value = task.exception?.message ?: "Login Gagal"
                }
            }
    }

    fun isUserLoggedIn(): Boolean = mAuth.currentUser != null

    fun logout() {
        mAuth.signOut()
    }

    fun loadAllObat() {
        _isLoading.value = true
        apiService.getAllObats().enqueue(object : Callback<ApiResponse<List<Obat>>> {
            override fun onResponse(
                call: Call<ApiResponse<List<Obat>>>,
                response: Response<ApiResponse<List<Obat>>>
            ) {
                _isLoading.value = false
                if (response.isSuccessful) {
                    _obatList.value = response.body()?.data
                } else {
                    _errorMessage.value = "Gagal memuat data obat"
                }
            }

            override fun onFailure(call: Call<ApiResponse<List<Obat>>>, t: Throwable) {
                _isLoading.value = false
                _errorMessage.value = "Error: ${t.message}"
            }
        })
    }

    fun loadObatByJenis(jenis: String) {
        _isLoading.value = true
        apiService.getObatByJenis(jenis).enqueue(object : Callback<ApiResponse<List<Obat>>> {
            override fun onResponse(
                call: Call<ApiResponse<List<Obat>>>,
                response: Response<ApiResponse<List<Obat>>>
            ) {
                _isLoading.value = false
                if (response.isSuccessful) {
                    _obatList.value = response.body()?.data
                } else {
                    _errorMessage.value = "Gagal memuat data obat"
                }
            }

            override fun onFailure(call: Call<ApiResponse<List<Obat>>>, t: Throwable) {
                _isLoading.value = false
                _errorMessage.value = "Error: ${t.message}"
            }
        })
    }

    fun getObatById(id: Int): LiveData<Obat> {
        val result = MutableLiveData<Obat>()
        apiService.getObatById(id).enqueue(object : Callback<ApiResponse<Obat>> {
            override fun onResponse(
                call: Call<ApiResponse<Obat>>,
                response: Response<ApiResponse<Obat>>
            ) {
                if (response.isSuccessful) {
                    result.value = response.body()?.data
                }
            }

            override fun onFailure(call: Call<ApiResponse<Obat>>, t: Throwable) {
                _errorMessage.value = "Error: ${t.message}"
            }
        })
        return result
    }

    fun deleteObat(id: Int) {
        apiService.deleteObat(id).enqueue(object : Callback<ApiResponse<String>> {
            override fun onResponse(
                call: Call<ApiResponse<String>>,
                response: Response<ApiResponse<String>>
            ) {
                if (response.isSuccessful) {
                    loadAllObat()
                }
            }

            override fun onFailure(call: Call<ApiResponse<String>>, t: Throwable) {
                _errorMessage.value = "Error: ${t.message}"
            }
        })
    }

    fun updateStock(id: Int, newStock: Int) {
        apiService.updateStock(id, UpdateStockRequest(newStock))
            .enqueue(object : Callback<ApiResponse<Obat>> {
                override fun onResponse(
                    call: Call<ApiResponse<Obat>>,
                    response: Response<ApiResponse<Obat>>
                ) {
                    if (response.isSuccessful) {
                        loadAllObat()
                    }
                }

                override fun onFailure(call: Call<ApiResponse<Obat>>, t: Throwable) {
                    _errorMessage.value = "Error: ${t.message}"
                }
            })
    }

    fun loadAllSuppliers() {
        apiService.getAllSuppliers().enqueue(object : Callback<ApiResponse<List<Supplier>>> {
            override fun onResponse(
                call: Call<ApiResponse<List<Supplier>>>,
                response: Response<ApiResponse<List<Supplier>>>
            ) {
                if (response.isSuccessful) {
                    _supplierList.value = response.body()?.data
                }
            }

            override fun onFailure(call: Call<ApiResponse<List<Supplier>>>, t: Throwable) {
                _errorMessage.value = "Error: ${t.message}"
            }
        })
    }
}
